File: Lectures/11-Greedy-Algorithms/Efficient-Huffman-Coding.cpp
#include <bits/stdc++.h>
using namespace std;

struct node {
    int freq;
    string symbol;
    node* left;
    node* right;
    string huff;

    node(int freq=0, string symbol="", node* left=NULL, node* right=NULL, string huff="") {
        this->freq = freq;
        this->symbol = symbol;
        this->left = left;
        this->right = right;
        this->huff = huff;
    }
};

void print_nodes(node* curr, string val = "") {
    string new_val = val + curr->huff;
    if (curr->left)
        print_nodes(curr->left, new_val);
    if (curr->right)
        print_nodes(curr->right, new_val);
    if (!curr->left && !curr->right)
        cout << curr->symbol << " -> " << new_val << '\n';
}

int main() {
    int n = 6;
	string chars[] = {"a", "b", "c", "d", "e", "f"};
	int freq[]     = { 5,   9,   12,  13,  16, 45 };
	auto comp = [](auto& x, auto& y){return x->freq > y->freq;};
	priority_queue<node*, vector<node*>, decltype(comp)> nodes(comp);
	for (int i = 0; i < n; i++) {
        node* curr = new node(freq[i], chars[i]);
		nodes.push(curr);
    }
	while (nodes.size() > 1) {
		node* left = nodes.top();
		nodes.pop();
		node* right = nodes.top();
		nodes.pop();
		left->huff = "0";
		right->huff = "1";
		node* new_node = new node(left->freq+right->freq, left->symbol+right->symbol, left, right);
		nodes.push(new_node);
	}
	print_nodes(nodes.top());
}

File: Lectures/11-Greedy-Algorithms/Efficient-Huffman-Coding.py
import heapq

class node:
    def __init__(self, freq, symbol, left=None, right=None, huff=''):
        self.freq = freq
        self.symbol = symbol
        self.left = left
        self.right = right
        self.huff = huff
    def __lt__(self, x):
        return x.freq > self.freq

def print_nodes(node, val = ''):
    new_val = val + str(node.huff)
    if (node.left):
        print_nodes(node.left, new_val)
    if (node.right):
        print_nodes(node.right, new_val)
    if (not node.left and not node.right):
        print(node.symbol, "->", new_val)
 
chars = ['a', 'b', 'c', 'd', 'e', 'f']
freq  = [ 5,   9,   12,  13,  16, 45 ]
nodes = []
for i in range(len(chars)):
    nodes.append(node(freq[i], chars[i]))
heapq.heapify(nodes)
while len(nodes) > 1:
    left = heapq.heappop(nodes)
    right = heapq.heappop(nodes)
    left.huff = 0
    right.huff = 1
    new_node = node(left.freq+right.freq, left.symbol+right.symbol, left, right)
    heapq.heappush(nodes, new_node)
print_nodes(heapq.heappop(nodes))
